### PR TITLE
[WIP] add wenetspeech recipe

### PR DESCRIPTION
**Pull request by pingfengluo**
Features:
- [x] prepare data and  make a char lexicon from training text
- [ ] train a RNN-T model (char-unit,and may add pinyin-unit/phone-unit  too)

this PR is same  as https://github.com/k2-fsa/icefall/pull/165 , just removed some logs.

### CODE REVIEW

**Comment by pingfengluo** [comment]
commands to run WenetSpeech recipe are :

```
cd icefall/egs/wenetspeech/ASR && ./prepare.sh  && ./transducer_stateless/train.py \
  --bucketing-sampler True \
  --world-size 8 \
  --lang-dir data/lang_char \
  --num-epochs 40 \
  --start-epoch 0 \
  --exp-dir transducer_stateless/exp_char \
  --max-duration 200 \
  --lr-factor 3

```

**Comment by pingfengluo** [comment]
> As I suggested before, you may want to try running the training loop with commented out code inside to confirm if it's related to dataloading or sth else.

CPU OOM  may be related to dataloading, because there still was CPU OOM when commented out other code.

Would show the full stack trace next time when  hit CPU OOM.

File: egs/wenetspeech/ASR/local/compute_fbank_wenetspeech.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+                # precompute features on shorter chunks,
+                # without memory blow-ups.
+                if torch.cuda.is_available():
+                    logging.info("GPU detected, do the CUDA extraction.")
+                    cut_set = cut_set.compute_and_store_features_batch(
+                        extractor=extractor,
+                        storage_path=f"{output_dir}/feats_{partition}",
+                        num_workers=args.num_workers,
+                        batch_duration=args.batch_duration,
+                        storage_type=LilcomHdf5Writer,
```
[the remark this answers is not in the document]
Comment (pingfengluo) [reply egs/wenetspeech/ASR/local/compute_fbank_wenetspeech.py:237]: > If you're not using on-the-fly feature extraction, then the growing memory usage is likely due to using HDF5 here (and in other places). You can try replacing with `LilcomChunkyWriter` which is available in this PR [lhotse-speech/lhotse#522](https://github.com/lhotse-speech/lhotse/pull/522) (it seems stable to me but I want to run it some more before I merge to be 100% sure).

will try it

[the remark this answers is not in the document]
Comment (pingfengluo) [reply egs/wenetspeech/ASR/local/compute_fbank_wenetspeech.py:237]: > 

@pzelasko 1. LilcomChunkyWriter in this PR https://github.com/lhotse-speech/lhotse/pull/522 is ok on AIShell and WenetSpeech. 2. But Lhotse filter takes too many menory  is normal (like https://github.com/k2-fsa/icefall/blob/df1e9919026b09145dbf148ca6a801eff66175a0/egs/wenetspeech/ASR/transducer_stateless/train.py#L600)?

[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/wenetspeech/ASR/local/compute_fbank_wenetspeech.py:237]: `train_cuts.filter()` will make a copy of `train_cuts` inside, which may increase noticeable CPU RAM usage if `train_cuts` is very large.

An alternative is to use
https://github.com/lhotse-speech/lhotse/blob/84f3d444b3ac15cb45d06872c9f5fb96b190ccfe/lhotse/dataset/sampling/base.py#L111

```python3
    def filter(self, predicate: Callable[[Cut], bool]) -> None:
        """
        Add a constraint on individual cuts that has to be satisfied to consider them.
        Can be useful when handling large, lazy manifests where it is not feasible to
        pre-filter them before instantiating the sampler.
        Example:
            >>> cuts = CutSet(...)
            ... sampler = SingleCutSampler(cuts, max_duration=100.0)
            ... # Retain only the cuts that have at least 1s and at most 20s duration.
            ... sampler.filter(lambda cut: 1.0 <= cut.duration <= 20.0)
        """
        self._filter_fn = predicate
```

It is used like
https://github.com/lhotse-speech/lhotse/blob/84f3d444b3ac15cb45d06872c9f5fb96b190ccfe/lhotse/dataset/sampling/single_cut.py#L212

```python3
            if not self._filter_fn(next_cut):
                # No - try another one.
                self.diagnostics.discard_single(next_cut)
                continue
```

Specifically, you can change
https://github.com/k2-fsa/icefall/blob/df1e9919026b09145dbf148ca6a801eff66175a0/egs/wenetspeech/ASR/tdnn_lstm_ctc/asr_datamodule.py#L286
to
```python3
 logging.info("About to create train dataloader") 

def remove_short_and_long_utt(c: Cut):
    # Keep only utterances with duration between 1 second and 20 seconds
    return 1.0 <= c.duration <= 20.0

train_sampler.filter(remove_short_and_long_utt)
```

[the remark this answers is not in the document]
Comment (pingfengluo) [reply egs/wenetspeech/ASR/local/compute_fbank_wenetspeech.py:237]: > 

I get it, thank you.

**Comment by pingfengluo** [comment]
when i tried RNNT_loss with K2 on WENET recipe, i got this error ( RNNT_loss with K2 on Librispeech is ok, the difference is   'blank_id: 2, vocab_size' 5537' ). Maybe you can help? @pkufool 
```
icefall/egs/wenetspeech/ASR# ./transducer_stateless/train.py   --bucketing-sampler True   --world-size 1   --lang-dir data/lang_char   --num-epochs 30   --start-epoch 0   --exp-dir transducer_stateless/exp_rnnt_k2   --max-duration 100   --lr-factor 3
2022-01-19 14:26:44,245 INFO [train.py:603] Training started
2022-01-19 14:26:44,252 INFO [train.py:613] Device: cuda:0
2022-01-19 14:26:44,266 INFO [lexicon.py:176] Loading pre-compiled data/lang_char/Linv.pt
2022-01-19 14:26:44,274 INFO [train.py:625] {'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_interval': 3000, 'log_diagnostics': False, 'feature_dim': 80, 'encoder_out_dim': 512, 'subsampling_factor': 4, 'attention_dim': 512, 'nhead': 8, 'dim_feedforward': 2048, 'num_encoder_layers': 12, 'vgg_frontend': False, 'warm_step': 1600000, 'env_info': {'k2-version': '1.11', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '5154032363e610d7b786a006a99a752c895d33da', 'k2-git-date': 'Mon Jan 17 12:11:52 2022', 'lhotse-version': '1.0.0.dev+git.c9f7643.clean', 'torch-cuda-available': True, 'torch-cuda-version': '10.2', 'python-version': '3.6', 'icefall-git-branch': 'master', 'icefall-git-sha1': '8f21e92-dirty', 'icefall-git-date': 'Fri Jan 14 12:12:38 2022', 'icefall-path': '/usr/local/lib/python3.6/dist-packages', 'k2-path': '/usr/local/lib/python3.6/dist-packages/k2/__init__.py', 'lhotse-path': '/usr/local/lib/python3.6/dist-packages/lhotse/__init__.py', 'hostname': 'notebook-10377-1640832887629-466767-rkj89.hy', 'IP address': '10.164.132.10'}, 'world_size': 1, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 30, 'start_epoch': 0, 'exp_dir': PosixPath('transducer_stateless/exp_rnnt_k2'), 'lang_dir': PosixPath('data/lang_char'), 'lr_factor': 3.0, 'context_size': 2, 'prune_range': 3, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 100, 'bucketing_sampler': True, 'num_buckets': 300, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'lazy_load': True, 'blank_id': 2, 'vocab_size': 5537}
2022-01-19 14:26:44,275 INFO [train.py:627] About to create model
2022-01-19 14:26:44,706 INFO [train.py:631] Number of model parameters: 89170849
2022-01-19 14:26:48,601 INFO [asr_datamodule.py:364] About to get train cuts
2022-01-19 14:26:48,601 INFO [asr_datamodule.py:366] use lazy cuts
2022-01-19 14:26:48,601 INFO [asr_datamodule.py:196] About to get Musan cuts
2022-01-19 14:26:50,022 INFO [asr_datamodule.py:203] Enable MUSAN
2022-01-19 14:26:50,023 INFO [asr_datamodule.py:228] Enable SpecAugment
2022-01-19 14:26:50,023 INFO [asr_datamodule.py:230] Time warp factor: 80
2022-01-19 14:26:50,023 INFO [asr_datamodule.py:244] About to create train dataset
2022-01-19 14:26:50,023 INFO [asr_datamodule.py:272] Using DynamicBucketingSampler.
2022-01-19 14:26:51,683 INFO [asr_datamodule.py:287] About to create train dataloader
2022-01-19 14:26:51,685 INFO [asr_datamodule.py:378] About to get dev cuts
2022-01-19 14:26:52,503 INFO [asr_datamodule.py:313] About to create dev dataset
/usr/local/lib/python3.6/dist-packages/lhotse/dataset/sampling/dynamic_bucketing.py:109: UserWarning: You are using DynamicBucketingSampler with an eagerly read CutSet. You won't see any memory/speed benefits with that setup. Either use 'CutSet.from_jsonl_lazy' to read the CutSet lazily, or use a BucketingSampler instead.
  "You are using DynamicBucketingSampler with an eagerly read CutSet. "
2022-01-19 14:26:52,510 INFO [asr_datamodule.py:332] About to create dev dataloader
2022-01-19 14:26:52,510 INFO [train.py:669] epoch 0, learning rate 0
/workspace/cpfs-data/code/icefall/egs/wenetspeech/ASR/transducer_stateless/conformer.py:113: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  lengths = ((x_lens - 1) // 2 - 1) // 2
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [0,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [1,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [2,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [3,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [4,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [5,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [6,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [7,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [8,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [9,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [10,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [11,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [12,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [13,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [14,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [15,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [16,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [17,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [18,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [19,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [20,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [21,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [22,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [279,0,0], thread: [23,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [0,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [1,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [2,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [3,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [4,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [5,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [6,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [7,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [8,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [9,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [10,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [11,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [12,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [13,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [14,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [15,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [16,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [17,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [18,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [19,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [20,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [21,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [22,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [23,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [24,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [25,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [26,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [27,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [28,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [29,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [30,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:111: operator(): block: [268,0,0], thread: [31,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
Traceback (most recent call last):
  File "./transducer_stateless/train.py", line 717, in <module>
    main()
  File "./transducer_stateless/train.py", line 710, in main
    run(rank=0, world_size=1, args=args)
  File "./transducer_stateless/train.py", line 681, in run
    world_size=world_size,
  File "./transducer_stateless/train.py", line 527, in train_one_epoch
    is_training=True,
  File "./transducer_stateless/train.py", line 393, in compute_loss
    simple_loss, pruned_loss = model(x=feature, x_lens=feature_lens, y=y)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/workspace/cpfs-data/code/icefall/egs/wenetspeech/ASR/transducer_stateless/model.py", line 113, in forward
    decoder_out, encoder_out, y_padded, blank_id, boundary, True
  File "/usr/local/lib/python3.6/dist-packages/k2/rnnt_loss.py", line 211, in rnnt_loss_simple
    px, py = get_rnnt_logprobs(lm, am, symbols, termination_symbol, boundary)
  File "/usr/local/lib/python3.6/dist-packages/k2/rnnt_loss.py", line 150, in get_rnnt_logprobs
    torch.tensor(float("-inf"), dtype=px_am.dtype, device=px_am.device),
RuntimeError: CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

**Comment by pkufool** [comment]
@pingfengluo  I suspect there may be a mismatch of you `max_symbol_id` and tensor dimension, try to print out the shape of am & lm tensor (i.e. the output of encoder and prediction network), the shape of am should be [B, T, C] and lm [B, S+1, C], 
check `C` is greater or equal to `max_sym_id + 1`.

**Comment by pingfengluo** [comment]
> @pingfengluo I suspect there may be a mismatch of you `max_symbol_id` and tensor dimension, try to print out the shape of am & lm tensor (i.e. the output of encoder and prediction network), the shape of am should be [B, T, C] and lm [B, S+1, C], check `C` is greater or equal to `max_sym_id + 1`.
> 
> [edit]: You can get `max_sym_id` by `torch.max(y_padded)`, normally should be `vocab_size - 1`.

I just reduce vocab_size from '5537' to '504'  and the training script is running ok now. @pkufool

**Comment by pkufool** [comment]
Ok, I think the vocab-size should not make a difference.  Can you check the info I mention above (using vocab-size 5537), let's see if there is a mismatch.

**Comment by pingfengluo** [comment]
> Ok, I think the vocab-size should not make a difference. Can you check the info I mention above (using vocab-size 5537), let's see if there is a mismatch.

You are right, there is a 'mismatch of you max_symbol_id and tensor dimension'. How to fix it? @pkufool 
```
Traceback (most recent call last):
  File "./transducer_stateless/train.py", line 716, in <module>
    main()
  File "./transducer_stateless/train.py", line 709, in main
    run(rank=0, world_size=1, args=args)
  File "./transducer_stateless/train.py", line 680, in run
    world_size=world_size,
  File "./transducer_stateless/train.py", line 526, in train_one_epoch
    is_training=True,
  File "./transducer_stateless/train.py", line 392, in compute_loss
    simple_loss, pruned_loss = model(x=feature, x_lens=feature_lens, y=y)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/workspace/cpfs-data/code/icefall/egs/wenetspeech/ASR/transducer_stateless/model.py", line 113, in forward
    [encoder_out.size(), max_sym_id]
AssertionError: [torch.Size([15, 156, 512]), tensor(4135, device='cuda:0')]
```

**Comment by pingfengluo** [comment]
> @pingfengluo I suspect there may be a mismatch of you `max_symbol_id` and tensor dimension, try to print out the shape of am & lm tensor (i.e. the output of encoder and prediction network), the shape of am should be [B, T, C] and lm [B, S+1, C], check `C` is greater or equal to `max_sym_id + 1`.
> 
> [edit]: You can get `max_sym_id` by `torch.max(y_padded)`, normally should be `vocab_size - 1`.

Should `C` must be greater or equal to `max_sym_id + 1` ? I find that `C` is smaller than `max_sym_id + 1` and the training process is working ok, which is using torch rnnt_loss as https://github.com/k2-fsa/icefall/blob/master/egs/aishell/ASR/transducer_stateless/model.py

**Comment by pkufool** [comment]
@pingfengluo I think so,  C (i.e. am shape [B, T, C]) must be greater or equal to `vocab_size - 1` in k2 pruned rnnt loss. The training using torch rnnt_loss is ok, because it uses only the output of the joiner network, there is a linear layer to transform the dim to vocab-size. In k2's pruned rnnt loss, we use the output of encoder network and predict network to get the prune ranges, so we have to make some changes. The simplest way is to change the `encoder_out_dim` to vocab_size.

**Comment by pkufool** [comment]
@pingfengluo  BTW， The way you are using to get the blank id is not right, 2 is the id of oov (i.e. <unk>), just set blank_id to 0, you can see the id in the token.txt.

**Comment by pingfengluo** [comment]
> @pingfengluo BTW, the way you are using to get the blank id is not right, 2 is the id of oov (i.e. <unk>), just set blank_id to 0, you can see the id in the token.txt.

Yes, you are right.I will check it. thank you.

**Comment by pingfengluo** [comment]
> @pingfengluo I think so, C (i.e. am shape [B, T, C]) must be greater or equal to `vocab_size ` in k2 pruned rnnt loss. The training using torch rnnt_loss is ok, because it uses only the output of the joiner network, there is a linear layer to transform the dim to vocab-size. In k2's pruned rnnt loss, we use the output of encoder network and predict network to get the prune ranges, so we have to make some changes. The simplest way is to change the `encoder_out_dim` to vocab_size.
> 
> [edit]: The unpruned version of rnnt_loss in k2 has no such requirements.

@pkufool 
1. when i  changed the `encoder_out_dim` to  `vocab_size`, the model parameters increased from `89170849 ` to `123070912`, is about 50%.
2. The unpruned version of rnnt_loss in k2 is slower than the pruned version, or it takes less GPU memory?

**Comment by pkufool** [comment]
Yes, the parameters of several components would increase, like embedding layer in decoder, output_layer in encoder, linear layer in joiner. I just said that changing encoder_out_dim to vocab_size could fix the error,  obviously, there should be other better ways.  I will push one soon.

**Comment by pingfengluo** [comment]
@pzelasko `lhotse split` command takes too many CPU RAM on 10K hour dataset, if there is any solution to reduce RAM usage?

**Comment by pzelasko** [comment]
Ouch, it seems it's an "eager" split. Can you try an iterative version like this one? Sorry don't have time right now to update lhotse properly, will do it later.

```python
from math import ceil
from lhotse import load_manifest_lazy, CutSet

cuts = load_manifest_lazy("...jsonl.gz")
num_cuts = len(cuts)  # this counts newlines in JSONL quite quickly for lazy manifests
num_splits = 30
split_size = ceil(num_cuts / num_splits)

split = []
idx = 0
for cut in cuts:
    split.append(cut)
    idx += 1
    if idx == split_size:
        CutSet.from_cuts(split).to_file("...jsonl.gz")
        split = []
        idx = 0

if split:
    CutSet.from_cuts(split).to_file("...jsonl.gz")
```

**Comment by pzelasko** [comment]
@pkufool FYI https://github.com/lhotse-speech/lhotse/pull/558

**Comment by danpovey** [comment]
> update a CER result on WenetSpeech so far
> 
> model	TEST_NET	TEST_MEETING
> K2 RNNT	8.71	17.08

Great!
Can you comment about whether that's a good result?  How does it compare with other results on this dataset?

**Pull request closed without merging**